### PR TITLE
[Prototype, no review required]: mlflow.evaluate for LLM

### PR DESCRIPTION
**Pull request by harupy**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

(Please fill in changes proposed in this fix)

## How is this patch tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/models/evaluation/default_evaluator.py [hunk 1]
```diff
@@ -1316,3 +1316,65 @@ def get_original(self):
...
+        **kwargs,
+    ):
+        import evaluate
+
+        em = evaluate.load("exact_match")
+        df = dataset.features_data
+        preds = model.predict(df)
+        metrics = em.compute(references=dataset.labels_data, predictions=preds)
+        print(preds, metrics)
+        mlflow.llm.log_predictions(inputs=df.to_dict("records"), outputs=preds, prompts=preds)
```
Comment (dbczumar) [inline mlflow/models/evaluation/default_evaluator.py:1379]: Fetching "prompts" depends on the model type. For example, LangChain supports a flag called intermediate_outputs=True. We'll change the argument name of `mlflow.llm.log_predictions()` accordingly

File: example.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        model=model_info.model_uri,
+        data=df,
+        targets="target",
+        # New model types:
+        # - 'summarization'
+        # - 'qa'
+        # - 'retrieval'
+        # - 'text-generation'
+        model_type="qa",
+        evaluators=["qa"],
```
Comment (harupy) [inline example.py:86]: Determine what metrics needs to be computed for each model type.

File: example.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        model=model_info.model_uri,
+        data=df,
+        targets="target",
+        # New model types:
+        # - 'summarization'
+        # - 'qa'
+        # - 'retrieval'
+        # - 'text-generation'
+        model_type="text-generation",
+        evaluators=["text-generation"],
```
Comment (harupy) [inline example.py:42]:
```suggestion
```

Used `text-generation` for quick code update and separating evaluator classes, but should be removed and update the DefaultEvalutor.

File: mlflow/models/evaluation/default_evaluator.py [hunk 4]
```diff
@@ -1316,3 +1316,65 @@ def get_original(self):
...
+        preds = model.predict(df)
+        # How to log non-scalar metrics?
+        metrics = t.compute(predictions=preds)  # e.g. {'toxicity': [0.1, 0.2]}
+        print(preds, metrics)
+        # Should llm_predictions.csv contain metrics as well?
+        mlflow.llm.log_predictions(
+            inputs=df.to_dict("records"),
+            outputs=preds,
+            # How to get the prompts?
+            prompts=preds,
```
Comment (harupy) [inline mlflow/models/evaluation/default_evaluator.py:1351]:
```suggestion
            prompts=preds,
            metrics=list[{"toxicity": 0.1}, {"toxicity": 0.2}]
```

File: mlflow/models/evaluation/default_evaluator.py [hunk 5]
```diff
@@ -1316,3 +1316,65 @@ def get_original(self):
...
+        baseline_model=None,
+        **kwargs,
+    ):
+        import evaluate
+
+        t = evaluate.load("toxicity")
+        df = dataset.features_data
+        preds = model.predict(df)
+        # How to log non-scalar metrics?
+        metrics = t.compute(predictions=preds)  # e.g. {'toxicity': [0.1, 0.2]}
```
[the remark this answers is not in the document]
Comment (harupy) [reply mlflow/models/evaluation/default_evaluator.py:1344]: 1. Very toxic -> 1
2. Not toxic at all -> 0

(1 + 0) / 2 = 0.5

File: mlflow/models/evaluation/default_evaluator.py (hunk 4 above)
Comment (harupy) [reply mlflow/models/evaluation/default_evaluator.py:1351]: outputs is list[dict[str, Any]]?

{"output_text": "...", "metric1": ..., metric2: "..."}

Comment (harupy) [reply mlflow/models/evaluation/default_evaluator.py:1351]: mlflow.llm.log_predictions(
            inputs=df.to_dict("records"),
            outputs=preds,
            metadata=[{...}, {...}],  # -> ask Sunish

**Pull request closed without merging**